numba/cpython/builtins.py

```diff
@@ -751,7 +751,8 @@ def false_impl(var, typs):
         if isinstance(typ, types.TypeRef):
             # Use of Numba type classes is in general not supported as they do
             # not work when the jit is disabled.
-            if key not in (types.ListType, types.DictType):
+            if key not in (types.ListType, types.DictType, types.Tuple,
+                           types.UniTuple):
```

Comment:
Is this change actually demonstrating a potential bug in the impl that went out in 0.55 ?

Reply:
Thanks for confirming, indeed this refuses to compile in  0.55!

```python
from numba import njit, types
import numpy as np

@njit
def foo(x):
    if isinstance(x, types.UniTuple):
        print("is tuple")
    else:
        print("is NOT tuple")

foo((1,2))
foo(1)
```
The note above RE JIT transparency suggests that it ought to not work either?

>            # Use of Numba type classes is in general not supported as they do
>            # not work when the jit is disabled.

Reply:
the problem with using `isinstance(x, tuple)` is that it is hard to differentiate between an empty tuple and a non-empty one.

Reply:
I've update `rewrite_semantic_constants` to recognize non-arg tuples. This change allows the JIT to recognize and optimize away cases like `isisntance(x, tuple) and len(x) == 0`. Let's see if I didn't break Numba with this change